homeassistant/components/xiaomi_miio/config_flow.py

```diff
@@ -47,35 +51,73 @@ async def async_step_user(self, user_input=None):
             step_id="user", data_schema=CONFIG_SCHEMA, errors=errors
         )
 
+    async def async_step_zeroconf(self, user_input=None):
+        """Handle zeroconf discovery."""
+        if user_input is None:
+            return self.async_abort(reason="not_xiaomi_miio")
+
+        name = user_input.get("name")
+        self.host = user_input.get("host")
+        mac_address = user_input.get("properties", {}).get("mac")
+
+        if not name or not self.host or not mac_address:
+            return self.async_abort(reason="not_xiaomi_miio")
+
+        # Check which device is discovered.
+        if name.startswith(ZEROCONF_GATEWAY):
+            model = name.split("_")[0].replace("-", ".")
+
+            unique_id = f"{model}-{mac_address}-gateway"
```

Comment:
why is model part of unique ID? Can it be that a single device exposes 2 models on their mac address ?

Reply:
No that is not possible, you are right, the mac address would be enough for the unique ID. However the dev docs stated that models and mac are recommended for the unique ID, so at the time I implemented the Config Flow (manual init by user) I thought it was a good idea to use both model and mac because I can easily get the model once a connection is established.

The problem now is that when the device is discovered with zero conf I am missing the device token for authentication and I can't communicate with the device before we have got input from the user, but I already need to set and check for the unique ID.
Fortunately the first part of the name of the device is the model, at least with my gateway model.

I guess that the unique ID of zeroconf and manual initialized config flow need to be the same?
Am I allowed to remove the model from the already merged manual config flow?
(Wouldn't that cause problems for existing users, although this component is still only in dev merged and not yet in the release)

Reply:
You can make sure you update the existing entries during their set up.

We do something similar for Hue https://github.com/home-assistant/core/blob/dev/homeassistant/components/hue/__init__.py#L125-L129 . In your case you just test if unique ID ends with `-gateway` and if so, set mac address to be unique ID.

Reply:
alright this is done, it also has backwards compatibility and I have tested the migration and it works